File: _static/samples/sample.c
#include "id3Face.h"

void activateLicense()
{
    // [license_activation]
    int err;
    char hwCode[256];
    int hwCodeSize = 256;

    err = id3FaceLicense_GetHostHardwareCode(id3FaceLicenseHardwareCodeType_WindowsOs, hwCode, &hwCodeSize, NULL);

    if (err == id3FaceError_Success)
    {
        int licenseBufferSize = 1024*1024;
        unsigned char *licenseBuffer = (unsigned char*)malloc(size);

        err = id3FaceLicense_ActivateSerialKeyBuffer(hwCode, "XXXX-XXXX-XXXX-XXXX", "[Computer name]", licenseBuffer, &licenseBufferSize);
    }
    // [license_activation]
}

void checkLicense()
{
    // [license_check]
    int err;

    err = id3FaceLibrary_CheckLicense("id3Face.lic", NULL);

    if (err == id3FaceError_Success)
    {
    }
    // [license_check]
}

void checkLicenseBuffer()
{
    // [license_check_buffer]
    int err;

    err = id3FaceLibrary_CheckLicense(licenseBuffer, licenseBufferSize, NULL);

    if (err == id3FaceError_Success)
    {
    }
    // [license_check_buffer]
}

int FaceDetection()
{
    // [face_detection]
    ID3_FACE_IMAGE image = NULL;
    ID3_FACE_DETECTOR detector = NULL;
    ID3_DETECTED_FACE_LIST detected_face_list = NULL;
    ID3_DETECTED_FACE detected_face = NULL;
    int face_count = 0;
    int err;

    // Load the face detection AI model.
    err = id3FaceLibrary_LoadModel(models_dir.c_str(), id3FaceModel_FaceDetector3B, id3FaceProcessingUnit_Cpu);

    // Create a new instance of the FaceDetector class.
    if (err == id3FaceError_Success)
    {
        err = id3FaceDetector_Initialize(&detector);

        if (err == id3FaceError_Success)
            err = id3FaceDetector_SetConfidenceThreshold(detector, 50);
            
        if (err == id3FaceError_Success)
            err = id3FaceDetector_SetModel(detector, id3FaceModel_FaceDetector3B);

        if (err == id3FaceError_Success)
            err = id3FaceDetector_SetNmsIouThreshold(detector, 40);

        if (err == id3FaceError_Success)
            err = id3FaceDetector_SetThreadCount(detector, 4);
    }

    // load an image from a file
    if (err == id3FaceError_Success)
    {
        err = id3FaceImage_Initialize(&image);
    
        if (err == id3FaceError_Success)
            err = id3FaceImage_FromFile(image, "image1.jpg", id3FacePixelFormat_Bgr24Bits);
    }

    // Detect faces on the image.
    if (err == id3FaceError_Success)
    {
        err = id3DetectedFaceList_Initialize(&detected_face_list);

        if (err == id3FaceError_Success)
            err = id3DetectedFace_Initialize(&detected_face);

        if (err == id3FaceError_Success)
            err = id3FaceDetector_DetectFaces(detector, image, detected_face_list);
    }   

    // Enumerate detected faces.
    if (err == id3FaceError_Success)
    {
        err = id3DetectedFaceList_GetCount(detected_face_list, &face_count);

        if (err == id3FaceError_Success)
        {
            for (int i = 0; i < face_count; i++)
            {
                err = id3DetectedFaceList_Get(detected_face_list, i, detected_face); 

                if (err == id3FaceError_Success)
                {
                    id3FaceRectangle bounds;
                    err = id3DetectedFace_GetBounds(detected_face, &bounds);
                }
            }
        }
    }

    // Disposes all resources
    if (detected_face != NULL)    
        id3DetectedFace_Dispose(&detected_face);

    if (detected_face_list != NULL)
        id3FaceDetectedFaceList_Dispose(&detected_face_list);

    if (detector != NULL)
        id3FaceDetector_Dispose(&detector);
    //[face_detection]

    return err;
}

void FaceComparisonOnCard()
{
    //[face_comparison_on_card]
    
    //[face_comparison_on_card]
}

void FaceAnalysis()
{
    ID3FACE_ANALYSER hAnalyser;
    int err;

    // Initialize the FaceAnalyser module
    err = id3FaceAnalyser_Initialize(&hAnalyser);

    if (err == id3FaceError_Success)
    {
        id3FaceAnalyser_SetExpressionSensibility(hAnalyser, 60);
        id3FaceAnalyser_SetOverExposureSensibility(hAnalyser, 70);
        id3FaceAnalyser_SetUnderExposureSensibility(hAnalyser, 70);

        // ...

        // Release all resources allocated to the FaceAnalyser module.
        err = id3FaceAnalyser_Dispose(hAnalyser);
    }
}

int main(const int argc, const char **argv)
{

    return 0;
}

File: _static/samples/sample.cs
using System;

namespace id3.Face.Samples
{
    using id3.Face;

    class Program
    {
        static void Main(string[] args)
        {
            string licensePath = @"id3FaceToolkit.lic";
            string modelPath = "../../../../../sdk/models";

            // [check_license]
            try
            {
                FaceLibrary.CheckLicense(licensePath);
            }
            catch (FaceException ex)
            {
                Console.WriteLine("Error during license check" + ex.Message);
                Environment.Exit(-1);
            }
            // [check_license]

            // [load_models]
            FaceLibrary.LoadModel(modelPath, FaceModel.FaceDetector3B, ProcessingUnit.Cpu);
            // [load_models]

            string imagePath1 = "../../../../../data/image1.jpg";
            string imagePath2 = "../../../../../data/image2.jpg";

            // [load_image]
            Image image1 = Image.FromFile(imagePath1, PixelFormat.Bgr24Bits);
            // [load_image]

            Image image2 = Image.FromFile(imagePath2, PixelFormat.Bgr24Bits);

            // [init_face_detector]
            FaceDetector faceDetector = new FaceDetector()
            {
                ConfidenceThreshold = 50,
                Model = FaceModel.FaceDetector3B,
                ThreadCount = 4
            };
            // [init_face_detector]

            DetectedFaceList detectedFaceList1 = faceDetector.DetectFaces(image1);
            DetectedFaceList detectedFaceList2 = faceDetector.DetectFaces(image2);

            // [init_face_encoder]
            FaceEncoder faceEncoder = new FaceEncoder()
            {
                Model = FaceModel.FaceEncoder9A,
                ThreadCount = 4
            };
            // [init_face_encoder]

            // [face_encoder_create_template]
            FaceTemplate faceTemplate1 = faceEncoder.CreateTemplate(image1, detectedFaceList1.GetLargestFace());
            // [face_encoder_create_template]

            FaceTemplate faceTemplate2 = faceEncoder.CreateTemplate(image2, detectedFaceList2.GetLargestFace());

            // [init_face_matcher]
            FaceMatcher faceMatcher = new FaceMatcher();
            // [init_face_matcher]

            int score = faceMatcher.CompareTemplates(faceTemplate2, faceTemplate1);

            if (score > (int)FaceMatcherThreshold.Fmr10000)
            {
                Console.WriteLine("Match: " + score);
            }
            else
            {
                Console.WriteLine("No match: " + score);
            }

            /**
	         * Face templates can be exported directly into a file or a buffer.
	         * When using the SDK face matcher the id3FaceTemplateBufferType_Normal must be used.
	        */
            Console.Write("Export template 1 as file...");
            faceTemplate1.ToFile(FaceTemplateBufferType.Normal, "../../../../../data/template1.bin");
            Console.Write("Export template 2 as buffer...");
            byte[] template2Buffer = faceTemplate2.ToBuffer(FaceTemplateBufferType.Normal);

            Console.WriteLine("Sample terminated successfully.");
            Console.ReadKey();
        }

        void CheckLicense()
        {
            // [license_check]
            try
            {
                FaceLibrary.CheckLicense("id3Face.lic");
            }
            catch (FaceException ex)
            {
                Console.WriteLine(ex.Message);
            }
            // [license_check]
        }
        void ActivateLicense()
        {
            // [license_activation]
            string hardwareCode = License.GetHostHardwareCode(LicenseHardwareCodeType.WindowsOs);
            
            try
            {
                License.ActivateSerialKey(hardwareCode, "XXXX-XXXX-XXXX-XXXX", "[Computer name]", "id3Face.lic");
            }
            catch (FaceException ex)
            {
                Console.WriteLine(ex.Message);
            }
            // [license_activation]
        }

        void FaceDetection()
        {
            // [face_detection]
            // Load the face detection AI model.
            FaceLibrary.LoadModel(modelPath, FaceModel.FaceDetector4B, ProcessingUnit.Cpu);

            // Create a new instance of the FaceDetector class.
            var faceDetector = new FaceDetector()
            {
                ConfidenceThreshold = 50,
                Model = FaceModel.FaceDetector4B,
                NmsIouThreshold = 40,
                ThreadCount = 4
            };

            // Load an image from a file.
            Image image = Image.FromFile("image1.jpg", PixelFormat.Bgr24Bits);

            // Detect faces on the image.
            DetectedFaceList detectedFaceList = faceDetector.DetectFaces(image);

            // Enumerate detected faces.
            foreach (DetectedFace face in detectedFaceList)
            {
                // ...                
            }

            // Disposes all resources allocated to the FaceDetector.
            faceDetector.Dispose();
            // [face_detection]
        }

        void FaceTracker()
        {
            // [face_tracking]
            // Load the face detection AI model.
            FaceLibrary.LoadModel(modelPath, FaceModel.FaceDetector4B, ProcessingUnit.Cpu);

            // Create a new instance of the FaceTracker class.
            var faceTracker = new FaceTracker()
            {
                ConfidenceThreshold = 50,
                Model = FaceModel.FaceDetector4B,
                NmsIouThreshold = 40,
                ThreadCount = 4
            };

            // Load an image from a file.
            Image image = Image.FromFile("image1.jpg", PixelFormat.Bgr24Bits);

            // Detect faces on the image.
            TrackedFaceList trackedFaceList = faceTracker.TrackFaces(image);

            // Enumerate tracked faces.
            foreach (TrackedFace face in trackedFaceList)
            {
                // ...                
            }

            // Dispose all resources allocated to the FaceTracker.
            faceTracker.Dispose();
            // [face_tracking]
        }

        void FaceComparisonOnCard()
        {
            // [face_comparison_on_card]
            FaceDetector faceDetector = new FaceDetector();
            FaceEncoder faceEncoder = new FaceEncoder();

            // Load an image from a file.
            Image image = Image.FromFile("image1.jpg", PixelFormat.Bgr24Bits);
           
            // Detect faces on the image.
            DetectedFaceList detectedFaceList = faceDetector.DetectFaces(image);

            // Create the template from the largest detected faces.
            FaceTemplate faceTemplate = faceEncoder.CreateTemplate(image, detectedFaceList.GetLargestFace());

            // Export the face template object to a Biometric Data Template (BDT) buffer.
            byte[] bdt = faceTemplate.ToBdt();

            // Send BDT buffer to card
            // ...

            // Dispose all resources
            detectedFaceList.Dispose();
            faceEncoder.Dispose();
            faceDetector.Dispose();
            // [face_comparison_on_card]
        }

        void FaceAnalysing()
        {
            // [face_analysing]
            FaceLibrary.LoadModel(ModelPath, FaceModel.FaceDetector3B, ProcessingUnit.Cpu);
            FaceLibrary.LoadModel(ModelPath, FaceModel.FaceAgeEstimator1A, ProcessingUnit.Cpu);
            FaceLibrary.LoadModel(ModelPath, FaceModel.FaceLandmarksEstimator2A, ProcessingUnit.Cpu);
            FaceLibrary.LoadModel(ModelPath, FaceModel.EyeGazeEstimator2A, ProcessingUnit.Cpu);

            // Initialize the FaceAnalyser module
            FaceAnalyser analyser = new FaceAnalyser()
            {
                ExpressionSensibility = 60,
                OverExposureSensibility = 70,
                UnderExposureSensibility = 70
            };

            // Initialize the FaceAnalyser module
            FaceAnalyser analyser = new FaceAnalyser();

            // Initialize the FaceDetector module
            FaceDetector faceDetector = new FaceDetector();

            // load image from file
            Image image = Image.FromFile("image.jpg", PixelFormat.Bgr24Bits);

            // detect face
            DetectedFaceList detectedFaceList = faceDetector.DetectFaces(image);
            DetectedFace detectedFace = detectedFaceList.GetLargestFace();

            // compute age
            int age = faceAnalyser.ComputeAge(image, detectedFace);

            // compute landmarks
            PointList landmarks = faceAnalyser.ComputeLandmarks(image, detectedFace);

            // compute eye-gaze
            EyeGaze eyeGaze = faceAnalyser.ComputeEyeGaze(image, detectedFace, landmarks);

            // Release all resources allocated to the FaceDetector module.
            faceDetector.Dispose();
 
            // Release all resources allocated to the FaceAnalyser module.
            analyser.Dispose();
            // [face_analysing]
        }
    }
}

